dashboard/config/database.yml

```diff
@@ -15,8 +15,30 @@ mysql_defaults: &mysql_defaults
   encoding: utf8
   collation: utf8_unicode_ci
   pool: 5
+
+  # number of seconds to wait for a connection to become available before giving up and raising a timeout error (default 5 seconds).
+  checkout_timeout: 3
+
+  # `mysql_options` client options forwarded through the mysql2 adapter:
+  # See: https://dev.mysql.com/doc/refman/5.7/en/mysql-options.html
+
+  # `MYSQL_OPT_CONNECT_TIMEOUT`: The connect timeout in seconds.
   connect_timeout: 2
+
+  # `MYSQL_OPT_RECONNECT`: Enable or disable automatic reconnection to the server if the connection is found to have
+  # been lost. Reconnect is off by default; this option provides a way to set reconnection behavior explicitly.
+  # See https://dev.mysql.com/doc/refman/5.7/en/c-api-auto-reconnect.html
   reconnect: true
+
+  # `MYSQL_OPT_READ_TIMEOUT`: The timeout in seconds for each attempt to read from the server.
+  # There are retries if necessary, so the total effective timeout value is three times the option value.
+  # You can set the value so that a lost connection can be detected earlier than the TCP/IP Close_Wait_Timeout value of 10 minutes.
+  read_timeout: 2
```

Comment:
What happens if a query takes longer than 2 seconds to execute on the server?  Does the SQL client and TCP stack consider that to exceed 2 seconds without reading anything on the socket?

Reply:
Good point- I ran a local test (with `select sleep(3);`) and it raised a `Mysql2::Error`. After further investigation it turns out that in addition to setting `mysql_options`, the `mysql2` adapter also uses the same variable to set the timeout on its own read, without any retry. So I think we should set this to be equivalent to our maximum expected query execution time, which I think is currently 30 seconds (as enforced by the server). I've updated the commit to reflect this.

Reply:
Will the handful of scheduled jobs that explicitly set a longer query execution timeout be impacted by the 30 second read timeout?  I assume they use the same mysql adapter and settings?  Is there the sql client equivalent of HTTP Keep-Alive?

Reply:
> Will the handful of scheduled jobs that explicitly set a longer query execution timeout be impacted by the 30 second read timeout? I assume they use the same mysql adapter and settings?

Yes, supporting the handful of long-running (analytical) queries that are reusing the same DB adapter as our live-site transactional queries poses a big problem for this PR, since the `read_timeout` setting needs to be set longer than any possible query executed with it, it's a client-adapter setting (not per-query), and supporting super-long queries and setting low timeouts for resiliency seem to be fundamentally incompatible with each other on the same adapter (e.g., our longest `MAX_EXECUTION_TIME` is currently 20 minutes which is far too long to wait for a database-failover).

Before I can merge this PR I'll need to do some more targeted refactoring to separate these two different types of queries from each other so they can use separate, differently-configured DB adapters to run their queries.

(Ideally, the two types of queries would be querying separate database-endpoints altogether with the long-running analytical queries going to the `reporting` DB, but if I recall there are specific reasons this refactoring either can't happen or hasn't happened yet.)

> Is there the sql client equivalent of HTTP Keep-Alive?

Yes and no. Yes, MySQL connections use a keep-alive mechanism to hang on to open TCP connections across queries for a specified duration (as configured by the [`wait_timeout`](https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html#sysvar_wait_timeout) / [`interactive_timeout`](https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html#sysvar_interactive_timeout) server system variables).

However, assuming you're asking instead whether there's a keepalive (or retry?) mechanism _for each individual query_, the answer to that is that for `mysql2` at least no, the adapter does not provide any such mechanism, [`mysql2/client.c`](https://github.com/brianmario/mysql2/blob/c578718cb9adc49117358162256420d977381060/ext/mysql2/client.c#L660) passes the `read_timeout` directly to the OS fd-read command as its timeout argument, and raises an exception if timeout occurs.